# Part 1. Equation of a Slime

How many late days are you using for this assignment? 2

In [16]:
# Imports section
import pandas as pd
import numpy as np
from sklearn.model_selection import cross_val_score, train_test_split, KFold
from sklearn.linear_model import LinearRegression, LogisticRegression
from sklearn.preprocessing import PolynomialFeatures, StandardScaler
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier

## 1. Loading the dataset

In [17]:
# Using pandas load the dataset
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)

data = pd.read_csv('science_data_large.csv')
print(data.head(15))
print(data.info())

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 n

## 2. Splitting the dataset

In [18]:
# Take the pandas dataset and split it into our features (X) and label (y)

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
# For grading consistency use random_state=42

X = data[['Temperature °C', 'Mols KCL']]  # Independent variables
y = data['Size nm^3']  # Dependent variable

# Split into training (90%) and testing (10%) sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)

# Output dataset shapes
print(f"Training Features: {X_train.shape}, Training Labels: {y_train.shape}")
print(f"Testing Features: {X_test.shape}, Testing Labels: {y_test.shape}")

Training Features: (900, 2), Training Labels: (900,)
Testing Features: (100, 2), Testing Labels: (100,)


## 3. Perform a Linear Regression

In [19]:
# Use sklearn to train a model on the training set

# Create a sample datapoint and predict the output of that sample with the trained model

# Report the score for that model using the default score function property of the SKLearn model, in your own words (markdown, not code) explain what the score means

# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX

linear = LinearRegression()
linear.fit(X_train, y_train)
sample = [[500.5,459.5]]
prediction = linear.predict(sample)


print(f"Predicted Size (nm³) for sample input: {prediction[0]:.5f}")

model_score = linear.score(X_test, y_test)
print(f"Model Score (R²): {model_score:.4f}")


coefficients = linear.coef_
intercept = linear.intercept_

print(f"Intercept: {intercept}")
print(f"Coefficients: {coefficients}")

Predicted Size (nm³) for sample input: 498638.18336
Model Score (R²): 0.8552
Intercept: -409391.4795834075
Coefficients: [ 866.14641337 1032.69506649]


/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


Write the linear equation of a slime: (example equation: $E = mc^2$):

### Linear Regression Equation:

The equation of our model is:

\[
$h(x) = -409391.47958 + (866.14641 \cdot x_1) + (1032.69507 \cdot x_2)$
\]

Where:
- \( $x_1$ \) represents **Temperature (°C)**
- \( $x_2$ \) represents **Mols KCl**
- \( $h(x)$ \) predicts the **Size (nm³)**

Report on score and explain meaning:

/Start 383GPT

### Model Score Interpretation (R² Score)

The R² score measures how well our regression model explains the variance in the target variable (`Size nm³`).  
- A score of **1.0** means the model **perfectly** predicts the output.
- A score of **0** means the model **performs no better than guessing the mean of the data**.
- A **negative score** suggests the model is worse than a constant baseline.

A higher R² value indicates that the input features (`Temperature °C` and `Mols KCl`) effectively explain the variation in slime size.

/End 383GPT

Since my value is **0.8552**, this means we are much closer to perfectly predicting the output rather than performing no better than guessing or worse.

## 4. Use Cross Validation

In [20]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42

# Report on their finding and their significance
kf = KFold(n_splits=5, shuffle=True, random_state=42)

# Perform cross-validation
cv_scores = cross_val_score(linear, X, y, cv=kf, scoring='r2')

# Print results
print(f"Cross-Validation R² Scores: {cv_scores}")
print(f"Mean R² Score: {np.mean(cv_scores):.4f}")
print(f"Standard Deviation of R²: {np.std(cv_scores):.4f}")

Cross-Validation R² Scores: [0.86151889 0.82742341 0.87195173 0.88166206 0.85609101]
Mean R² Score: 0.8597
Standard Deviation of R²: 0.0184


Write findings here:

/Start 383GPT

### Cross-Validation Results & Significance

Cross-validation helps assess model performance by training on different subsets of data. Here’s what our results indicate:

- **Mean R² Score**: Represents the model’s overall predictive power. A higher value suggests that the model explains more variance in slime size.
- **Standard Deviation of R²**: Measures consistency across different splits. A lower value means the model performs **similarly on different datasets**, which is a good sign.

If R² varies significantly across folds, it could indicate **overfitting** (model depends too much on specific data points) or **data variability** (some folds have more informative data than others).

/End 383GPT

Both our mean R^2 score and std deviation are good values, as mean R^2 is very close to 1 and also our R^2 score from the previous segment, and our values do not deviate much at all.


## 5. Using Polynomial Regression

In [21]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
# Perform k-fold cross validation (as above)

# Report on the metrics and output the resultant equation as you did in Part 3.
poly = PolynomialFeatures(degree=2)
X_poly = poly.fit_transform(X)

# Perform regression
model = LinearRegression()
model.fit(X_poly, y)


k = 5  
cv_scores = cross_val_score(model, X_poly, y, cv=k, scoring='neg_mean_squared_error')
mean_cv_score = -np.mean(cv_scores)


y_pred = model.predict(X_poly)
mse = mean_squared_error(y, y_pred)
r2 = r2_score(y, y_pred)

coefficients = model.coef_
intercept = model.intercept_
equation_terms = [f"{coeff:.3f}*{term}" for coeff, term in zip(coefficients, poly.get_feature_names_out())]
equation = f"{intercept:.3f} + " + " + ".join(equation_terms[1:])

print(f"Mean CV MSE: {mean_cv_score:.3f}")
print(f"Training MSE: {mse:.3f}")
print(f"R^2 Score: {r2:.3f}")
print(f"Resultant Equation: {equation}")


Mean CV MSE: 0.000
Training MSE: 0.000
R^2 Score: 1.000
Resultant Equation: 0.000 + 12.000*Temperature °C + -0.000*Mols KCL + -0.000*Temperature °C^2 + 2.000*Temperature °C Mols KCL + 0.029*Mols KCL^2


Write the polynomial equation of a slime: (example equation: $E = mc^2$)

Size = $14 \cdot$ Temperature °C + $0.029 \cdot$ Mols KCL^2

Report on the score and interpret:

The score is 1, which means it is predicting it perfectly. I'm not sure why this is the case and I believe it to be an error, however I have tried many changes and have not been able to see any results. It seems its a result of overfitting, which may be the intention here but I'm not sure. Regardless, I think we would expect our R^2 score to be around 0.85-0.95 without overfitting, again showing our predictions compared to the actual results.

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [22]:
data = pd.read_csv('ckd_feature_subset.csv')


X = data.drop('Target_ckd', axis=1)
y = data['Target_ckd']


scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)


models = {
    'Logistic Regression': LogisticRegression(random_state=42, max_iter=1000),
    'Support Vector Machine': SVC(random_state=42),
    'k-Nearest Neighbors': KNeighborsClassifier(),
    'Neural Network': MLPClassifier(random_state=42, max_iter=1000)
}


results = {}
for name, model in models.items():

    cv_scores = cross_val_score(model, X_scaled, y, cv=5)
    results[name] = {
        'mean_accuracy': cv_scores.mean(),
        'std_accuracy': cv_scores.std()
    }

print("Classification Model Performance:")
print("Model | Mean Accuracy | Standard Deviation")
print("------|--------------|-------------------")
for name, metrics in results.items():
    print(f"{name} | {metrics['mean_accuracy']:.4f} | {metrics['std_accuracy']:.4f}")

nn_configs = [
    {
        'hidden_layer_sizes': (10,),
        'activation': 'relu',
        'solver': 'adam'
    },
    {
        'hidden_layer_sizes': (50, 25),
        'activation': 'tanh',
        'solver': 'sgd'
    },
    {
        'hidden_layer_sizes': (100, 50, 25),
        'activation': 'logistic',
        'solver': 'lbfgs'
    }
]


Classification Model Performance:
Model | Mean Accuracy | Standard Deviation
------|--------------|-------------------
Logistic Regression | 0.9413 | 0.0240
Support Vector Machine | 0.9542 | 0.0339
k-Nearest Neighbors | 0.9411 | 0.0528
Neural Network | 0.9673 | 0.0211


## Results and Conclusion for Classification Experiments

In the classification analysis of Chronic Kidney Disease prediction, we evaluated four machine learning models using cross-validation. The models demonstrated very high performance, with accuracies ranging between 94% and 97%. The Neural Network emerged as the top performer, achieving a mean accuracy of 96.73% with the most consistent results across different data subsets. Support Vector Machines closely followed with a 95.42% accuracy.
The performance suggests that the indicators & their corrospondign values that are present in the dataset are accurate predictors of CKD, and therefore valuable to have as data points. The high accuracy across all models indicates that machine learning techniques can potentially provide powerful tools for early disease detection, offering clinicians a reliable method to screen for CDK using a small set of key physiological measurements through ML & AI.

In [23]:
nn_results = {}
for i, config in enumerate(nn_configs, 1):
    nn = MLPClassifier(
        hidden_layer_sizes=config['hidden_layer_sizes'],
        activation=config['activation'],
        solver=config['solver'],
        random_state=42,
        max_iter=1000
    )
    
    cv_scores = cross_val_score(nn, X_scaled, y, cv=5)
    nn_results[f'Config {i}'] = {
        'config': config,
        'mean_accuracy': cv_scores.mean(),
        'std_accuracy': cv_scores.std()
    }

print("\nNeural Network Configuration Experiments:")
print("Config | Hidden Layers | Activation | Solver | Mean Accuracy | Std Deviation")
print("-------|--------------|------------|--------|--------------|---------------")
for name, metrics in nn_results.items():
    config = metrics['config']
    print(f"{name} | {config['hidden_layer_sizes']} | {config['activation']} | {config['solver']} | {metrics['mean_accuracy']:.4f} | {metrics['std_accuracy']:.4f}")

/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  warnings.warn(
/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  warnings.warn(
/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  warnings.warn(



Neural Network Configuration Experiments:
Config | Hidden Layers | Activation | Solver | Mean Accuracy | Std Deviation
-------|--------------|------------|--------|--------------|---------------
Config 1 | (10,) | relu | adam | 0.9673 | 0.0211
Config 2 | (50, 25) | tanh | sgd | 0.9738 | 0.0248
Config 3 | (100, 50, 25) | logistic | lbfgs | 0.9802 | 0.0162


## Results and Conclusion for Neural Network Experiments

The neural network experiments, the configurations varied in hidden layer structure, activation functions, and optimization solvers, based on the advice of 383GPT. The most sophisticated model (again according to 383GPT), featuring three hidden layers with 100, 50, and 25 neurons and using the logistic activation function with the LBFGS solver, achieved the highest performance. This configuration reached 98.02% accuracy with the lowest variability across different data splits. The other configurations did almost just as well, with only a percent or two difference between them. These results highlight the importance of network architecture in machine learning models, demonstrating how increased complexity and carefully selected parameters can incrementally improve predictive performance for medical classification tasks.